File: S_MinimizationCNRotatingWP.m
%   S_MinimizationCNPRotatingWP.m
% Minimization of the CN Rotating WavePlate polarimeter
% El polarizador inicial se supone a lo largo del eje X
clear
clc
close all

warning off all
format compact
format long

Sini = [1,1,0,0]';

% Desfase lámina
philamina = pi/2;

%Genera esfera de Poincare
fact = 0.96; % radio ligeramente inferior a 1
[X, Y, Z] = sphere;
X = fact*X;
Y = fact*Y;
Z = fact*Z;

figure (2)
surf(X,Y,Z)
hold on
axis equal
xlabel('S1');
ylabel('S2');
zlabel('S3');
colormap('Bone');    
hold on


% Dibujo sobre la esfera del "8" de los estados de polarización generados
% al rotar la lámina

% Estados de polarización generados al rotar la lámina
Nrot = 100; % Número de rotaciones
SoPs = zeros(Nrot,4);

rot = linspace(0,pi,Nrot); % Rotaciones
for k=1:Nrot
    M = Mueller('waveplate', rot(k), philamina);
    SoPs(k,:) = M * Sini;
end

hold on
plot3(SoPs(:,2), SoPs(:,3), SoPs(:,4),'bo', 'MarkerFaceColor','b');
hold off

%% Minimización CN

% Numero de iteraciones
%===================
%===================
nitera = 400;
%===================
%===================

% Número de rotaciones de la lámina
Nrot = 6;

% Rotaciones iniciales
% ============================    
%     % Todos a cero
%     x0 = zeros(Nrot,1);
% 
    % aleatorios
    %=======================================
    x0 = pi*rand(Nrot,1);
    
    % "Equiespaciados"
    %=======================================
%     x0 = linspace(0.1, pi-0.1, Nrot);

    % Valor inicial del CN
    % valor inicial del Condition Number
CNB = RotatingWPPolCN(x0);
XB  = x0;  
    

% Minimización del CN a partir del punto inicial
%===============================================
% opciones para el programa de optimizacion
TolFun = 1.e-5; %1.e-15;
TolX   = 1.e-3; %1e-13;
options = optimset('MaxIter', 300, 'Display', 'iter', 'MaxFunEvals', 30000,'Algorithm','active-set','TolFun',TolFun,'TolX',TolX); 

% Las variables a optimizar en este caso son las rotaciones
% Intervalos donde se deja variar las rotaciones
% En este caso no my grande 
lb = 0; % Lower bounds
ub = pi; % Upper bounds
x = fmincon(@RotatingWPPolCN, x0, [], [], [], [], lb,ub, [], options);
x

[CN, TSoP] = RotatingWPPolSops(x, philamina);

if CN < CNB
    CNB = CN;
    XB   = x;
end

%Esto se dibuja en la figura 2
hold on
plot3(TSoP(:,2), TSoP(:,3), TSoP(:,4),'go', 'MarkerFaceColor','g');
hold off

% opciones para el programa de optimizacion
% Por si se quieren cambiar
options = optimset('MaxIter', 300, 'Display', 'off', 'MaxFunEvals', 30000, 'Algorithm','active-set','TolFun',TolFun); 

for n=1:nitera  % Bucle paralelo
    x0 = pi*rand(Nrot,1);
    x = fmincon(@RotatingWPPolCN, x0, [], [], [], [], lb,ub, [], options);

    [CN, TSoP] = RotatingWPPolSops(x, philamina);

    if CN < CNB
        CNB = CN;
        XB   = x;
    end
    disp(n)
    fprintf('CN=%e   CNB=%e',CN, CNB)
             
end

  
% Dibuja los estados de polarizacion en la figura 2
figure (2)
hold on
plot3(TSoP(:,2), TSoP(:,3), TSoP(:,4),'ro', 'MarkerFaceColor','r');
hold off

% Rotaciones en gados ordenadas
rotacion = sort(XB*180/pi)

File: S_MinimizationCNRotatingWP.py
import numpy as np
import matplotlib.pyplot as plt
from scipy.optimize import minimize
from mpl_toolkits.mplot3d import Axes3D  # noqa
from Mueller import Mueller
from RotatingWPPolCN import RotatingWPPolCN
from RotatingWSPPolSops import RotatingWPPolSops

# --------------------------------------------------
# Parámetros iniciales
# --------------------------------------------------

Sini = np.array([1, 1, 0, 0])
philamina = np.pi / 2

# --------------------------------------------------
# Esfera de Poincaré
# --------------------------------------------------

fact = 0.96
u = np.linspace(0, np.pi, 60)
v = np.linspace(0, 2*np.pi, 60)

X = fact * np.outer(np.sin(u), np.cos(v))
Y = fact * np.outer(np.sin(u), np.sin(v))
Z = fact * np.outer(np.cos(u), np.ones_like(v))

fig = plt.figure()
ax = fig.add_subplot(111, projection='3d')

ax.plot_surface(X, Y, Z, color='lightgray', alpha=0.5)

ax.set_xlabel('S1')
ax.set_ylabel('S2')
ax.set_zlabel('S3')

# --------------------------------------------------
# Estados de polarización (figura en 8)
# --------------------------------------------------

Nrot = 100
rot = np.linspace(0, np.pi, Nrot)

SoPs = np.zeros((Nrot, 4))

for k in range(Nrot):
    M = Mueller('waveplate', rot[k], philamina)
    SoPs[k, :] = M @ Sini

ax.scatter(SoPs[:, 1], SoPs[:, 2], SoPs[:, 3],
           color='blue', s=15)

# --------------------------------------------------
# Función objetivo (CN)
# --------------------------------------------------

def objective(x):
    return RotatingWPPolCN(x)

# --------------------------------------------------
# Optimización inicial
# --------------------------------------------------

Nrot_opt = 6

x0 = np.pi * np.random.rand(Nrot_opt)

bounds = [(0, np.pi)] * Nrot_opt

res = minimize(
    objective,
    x0,
    method='L-BFGS-B',
    bounds=bounds,
    options={'maxiter': 300}
)

x_best = res.x

CN_best, TSoP = RotatingWPPolSops(x_best, philamina)

# --------------------------------------------------
# Guardar mejor solución inicial
# --------------------------------------------------

CNB = CN_best
XB = x_best.copy()

# --------------------------------------------------
# Plot solución inicial
# --------------------------------------------------

ax.scatter(TSoP[:, 1], TSoP[:, 2], TSoP[:, 3],
           color='green', s=40)

# --------------------------------------------------
# Búsqueda global (equivalente al bucle MATLAB)
# --------------------------------------------------

nitera = 10

for n in range(nitera):

    x0 = np.pi * np.random.rand(Nrot_opt)

    res = minimize(
        objective,
        x0,
        method='L-BFGS-B',
        bounds=bounds,
        options={'maxiter': 300}
    )

    x = res.x
    CN, TSoP = RotatingWPPolSops(x, philamina)

    if CN < CNB:
        CNB = CN
        XB = x.copy()

    print(f"Iter {n+1} | CN = {CN:.6e} | Best CN = {CNB:.6e}")

# --------------------------------------------------
# Plot mejor solución
# --------------------------------------------------

ax.scatter(TSoP[:, 1], TSoP[:, 2], TSoP[:, 3],
           color='red', s=40)

plt.title("Poincaré Sphere - Optimization of CN")
plt.show()

# --------------------------------------------------
# Ángulos óptimos finales
# --------------------------------------------------

rotacion = np.sort(XB * 180 / np.pi)

print("\nOptimal angles (deg):")
print(rotacion)